mailbox/api/src/test/java/org/apache/james/mailbox/events/EventDeadLettersContract.java

```diff
@@ -0,0 +1,184 @@
+        assertThat(eventDeadLetters.groups().toStream())
+            .containsExactlyInAnyOrder(GROUP_A, GROUP_B);
+    }
+
+    @Test
+    default void groupsShouldReturnEmptyWhenNoStored() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        assertThat(eventDeadLetters.groups().toStream()).isEmpty();
+    }
+
+    @Test
+    default void storeShouldThrowWhenNullGroup() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        assertThatThrownBy(() -> eventDeadLetters.store(NULL_GROUP, EVENT_1))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    default void storeShouldThrowWhenNullEvent() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        assertThatThrownBy(() -> eventDeadLetters.store(GROUP_A, NULL_EVENT))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    default void storeShouldThrowWhenBothGroupAndEventAreNull() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        assertThatThrownBy(() -> eventDeadLetters.store(NULL_GROUP, NULL_EVENT))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    default void storeShouldStoreGroupWithCorrespondingEvent() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        assertThat(eventDeadLetters.take(GROUP_A).toStream())
+            .containsOnly(EVENT_1);
+    }
+
+    @Test
+    default void storeShouldWorkWhenMultipleDuplicatedStores() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+
+        assertThat(eventDeadLetters.take(GROUP_A).toStream())
+            .containsExactly(EVENT_1, EVENT_1, EVENT_1);
+    }
+
+    @Test
+    default void takeShouldThrowWhenGroupIsNull() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        assertThatThrownBy(() -> eventDeadLetters.take(NULL_GROUP))
+            .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    default void takeShouldReturnEventsRegisteredToGroup() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        eventDeadLetters.store(GROUP_A, EVENT_2).block();
+        eventDeadLetters.store(GROUP_B, EVENT_3).block();
+
+        assertThat(eventDeadLetters.take(GROUP_A).toStream())
+            .containsExactlyInAnyOrder(EVENT_1, EVENT_2);
+    }
+
+    @Test
+    default void takeShouldWipeOutStoredEventsCorrespondingToGroup() {
+        EventDeadLetters eventDeadLetters = eventDeadLetters();
+
+        eventDeadLetters.store(GROUP_A, EVENT_1).block();
+        eventDeadLetters.store(GROUP_A, EVENT_2).block();
+        eventDeadLetters.take(GROUP_A).then().block();
```

Comment:
Arg. IMO no. We should be able to read the events (think 'diagnostic': I want to see wich events did fail - not recompute them straight away.)

IMO sticking to the CRUD simple model is better: clearly separate read, writes & deletes.